source/common/router/scoped_config_impl.h

```diff
@@ -52,28 +55,36 @@ class ScopeKey {
   // Caller should guarantee the fragment is not nullptr.
   void addFragment(std::unique_ptr<ScopeKeyFragmentBase>&& fragment) {
     ASSERT(fragment != nullptr, "null fragment not allowed in ScopeKey.");
+    updateHash(*fragment);
     fragments_.emplace_back(std::move(fragment));
   }
 
+  uint64_t hash() const { return hash_; }
   bool operator!=(const ScopeKey& other) const;
-
   bool operator==(const ScopeKey& other) const;
 
 private:
+  // Update the key's hash with the new fragment hash.
+  void updateHash(const ScopeKeyFragmentBase& fragment) {
+    absl::uint128 buffer = absl::MakeUint128(hash_, fragment.hash());
+    hash_ = HashUtil::xxHash64(absl::string_view(reinterpret_cast<char*>(&buffer), 16));
```

Comment:
One observation that I think we need to start thinking more about independent of this is that we should move away from using non-cryptographic hashes for comparing configuration. I think that parts of the control plane should be viewed as untrusted, e.g. in a managed control plane, maybe route names are under the control of an adversary. In this scenario, it's plausible that hash collisions are not hard to create and we need to have something with strong crypto properties (at the expense of performance).
CC @mattklein123 @yanavlasov